src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+#include "postgres.h"
+#include "libpq-fe.h"
+#include "executor/spi.h"
+#include "distributed/lock_graph.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "postmaster/postmaster.h"
+#include "distributed/distribution_column.h"
+#include "utils/builtins.h"
+#include "distributed/shard_split.h"
+
+
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start);
+
+int MaxShardSize = 102400;
+
+/*
+ * Struct to store all the information related to
+ * a shard.
+ */
+typedef struct ShardInfoData
+{
+	int64 shardsize;
+	int64 shardminvalue;
+	int64 shardmaxvalue;
+	int64 shardid;
+	int32 nodeid;
+	char *tablename;
+	char *distributionColumn;
+	char *datatype;
+	char *shardname;
+	Oid tableId;
+	Oid distributionColumnId;
+}ShardInfoData;
+typedef ShardInfoData *ShardInfo;
+
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
+	if (HasNonTerminalJobOfType("rebalance", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("A rebalance is already running as job %ld", jobId),
+					errdetail("A rebalance was already scheduled as background job"),
+					errhint("To monitor progress, run: SELECT * FROM "
+							"citus_rebalance_status();")));
+	}
+	if (HasNonTerminalJobOfType("Automatic Shard Split", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("An automatic shard split is already running as job %ld",
+						   jobId),
+					errdetail(
+						"An automatic shard split was already scheduled as background job")));
+	}
+}
+
+
+/*
+ * For a given SplitPoints , it creates the SQL query for the shard Splitting
+ */
+StringInfo
+GetShardSplitQuery(ShardInfo shardinfo, List *SplitPoints , char* shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+
+	int64 length = list_length(SplitPoints);
+	appendStringInfo(splitQuery, "SELECT citus_split_shard_by_split_points(%ld, ARRAY[",
+					 shardinfo->shardid);
+
+	for (int i = 0; i < length - 1; i++)
+	{
+		appendStringInfo(splitQuery, "'%d',", list_nth_int(SplitPoints, i));
+	}
+	appendStringInfo(splitQuery, "'%d'], ARRAY[", list_nth_int(SplitPoints,
+																		  length - 1));
+
+	for (int i = 0; i < length; i++)
+	{
+		appendStringInfo(splitQuery, "%d,", shardinfo->nodeid);
+	}
+	appendStringInfo(splitQuery, "%d], %s)", shardinfo->nodeid, quote_literal_cstr(shardSplitMode));
+
+	return splitQuery;
+}
+
+
+/*
+ * It creates a background job for citus_split_shard_by_split_points and executes it in background.
+ */
+void
+ExecuteSplitBackgroundJob(int64 jobid, ShardInfo shardinfo, List *SplitPoints , char* shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+	splitQuery = GetShardSplitQuery(shardinfo, SplitPoints , shardSplitMode);
+	ereport(LOG, (errmsg(splitQuery->data)));
+	int32 nodesInvolved[1];
+	nodesInvolved[0] = shardinfo->nodeid;
```

Comment:
minor, can write this as `int32 nodesInvolved[] = { shardinfo->nodeid };`

Reply:
corrected